### PR TITLE
Add std.math/variance and std.math/std-dev

### PR DESCRIPTION
**Pull request by andrewberls**
This adds functions for computing the variance and standard deviation of
a collection of values, respectively. By default, these functions return
unbiased estimates, using a divisor of `n - 1` instead of `n` (Bessel's
Correction [0]). This is parameterizable by the optional `ddof`
parameter ("delta degreees of freedom") as in NumPy [1], where the divisor is `n -
ddof`; ddof = 0 will result in a biased estimator.

0: https://en.wikipedia.org/wiki/Bessel%27s_correction

1:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.var.html#numpy.var

### CODE REVIEW

File: src/framed/std/math.clj [hunk 1]
```diff
@@ -78,3 +78,28 @@
...
     v))
+
+(defn- squared-deviations
+  "Return the set of squared deviations from items in `vs` to their mean"
+  [mean vs]
+  (map #(Math/pow (- % mean) 2) vs))
+
+(defn variance
+  "Compute the variance of `vs`. Divisor used is `n - ddof` where `ddof` represents
+   'delta degrees of freedom'. By default returns an unbiased estimate (ddof = 1)
```
Comment (elliot42) [inline src/framed/std/math.clj]: Because this concept (both variance and delta degrees of freedom) is used but not explained here, I would recommend linking whatever definition/explanation was used in implementing this.  Some math concepts have multiple definitions and meanings so it's worth it to point to exactly what/why you mean.

File: src/framed/std/math.clj [hunk 2]
```diff
@@ -78,3 +78,28 @@
...
+  ([vs]
+   (variance vs 1))
+  ([vs ddof]
+   {:pre [(>= ddof 0)]}
+   (/ (reduce + (squared-deviations (mean vs) vs))
+      (- (count vs) ddof))))
+
+(defn std-dev
+  "Compute the standard deviation of `vs`. Divisor used is `n - ddof` where `ddof`
+   represents 'delta degrees of freedom'. By default returns an unbiased
```
Comment (elliot42) [inline src/framed/std/math.clj]: Ditto to link to undefined concepts as above.

File: test/framed/std/math_test.clj [hunk 3]
```diff
@@ -1,46 +1,62 @@
...
 (deftest test-mode
-  (is (= 2 (m/mode [1 2 3 2 3 2])))
-  (is (= [2 3] (m/mode [1 2 2 2 3 3 3])))
-  (is (nil? (m/mode [1 2 3 4 5]))))
+  (is (= 2 (std.math/mode [1 2 3 2 3 2])))
+  (is (= [2 3] (std.math/mode [1 2 2 2 3 3 3])))
+  (is (nil? (std.math/mode [1 2 3 4 5]))))
+
+(deftest test-variance
+  (let [vs [2 4 4 4 5 5 7 9]]
```
Comment (elliot42) [inline test/framed/std/math_test.clj]: I feel like this entire namespace is missing tests for empty lists, nil, mixed doubles and floats etc.  We don't need to fix everything up front but at least for these new functions I think we should start testing all the real conceivable inputs, e.g. empty lists, sequences as well as vectors, etc.

File: src/framed/std/math.clj [hunk 4]
```diff
@@ -78,3 +78,28 @@
...
+  (map #(Math/pow (- % mean) 2) vs))
+
+(defn variance
+  "Compute the variance of `vs`. Divisor used is `n - ddof` where `ddof` represents
+   'delta degrees of freedom'. By default returns an unbiased estimate (ddof = 1)
+   For a biased estimate, set ddof = 0"
+  ([vs]
+   (variance vs 1))
+  ([vs ddof]
+   {:pre [(>= ddof 0)]}
```
Comment (elliot42) [inline src/framed/std/math.clj]: I get the impression `ddof` also needs to be an integer?  y/n?

**Comment by elliot42** [comment]
This looks close but I'd like to see in the spec suite what is the expected behavior when these things are passed different variations on the conceivable input values.  Empty lists are definitely a super normal common case to encounter, so I think it should be explicitly/defined tested what the expected behavior is in that kind of case.

**Comment by elliot42** [comment]
Just in general with all specs ever you should expect to try to cover the total range of conceivable inputs and make sure the behavior of the function is defined logically, and actually behaviorally does the expected logical behavior.  You should think through these cases manually and/or lean on generative testing to generate more cases for you.

**Comment by andrewberls** [comment]
Great points. I'll add some more thorough test coverage here. Same for doc linking.

With regards to stuff like `ddof is an integer` - I would have thought your take on type assertions is that they are non-Clojure-y and that many clj functions take a "garbage-in-garbage-out" approach. I'm all for adding more assertions, but that was my thinking behind omitting it at the time.

**Comment by elliot42** [comment]
Thanks for checking the empty cases
